packages/hash/shared/src/types.ts

```diff
@@ -1,8 +1,46 @@
-// The types exported here are expected to be branded.
+import {
+  EntityId,
+  extractEntityUuidFromEntityId as extractEntityUuidFromEntityIdSubGraph,
+  extractOwnedByIdFromEntityId as extractOwnedByIdFromEntityIdSubGraph,
+  splitEntityId as splitEntityIdSubGraph,
+} from "@hashintel/hash-subgraph";
 
-export type Uuid = string;
-export type AccountId = Uuid;
-export type OwnedById = Uuid;
-export type UpdatedById = Uuid;
+type Branded<Base, Kind extends string> = {
+  /** The type of the value space that is branded */
+  readonly _base: Base;
+  /** The unique name for the branded type. */
+  readonly _kind: Kind;
+};
 
-export type EntityUuid = Uuid;
+/**
+ * The type-branding type to support nominal (name based) types.
+ */
+export type Brand<Base, Kind extends string> = Base & Branded<Base, Kind>;
+
+/** An account ID of an actor that is the owner of something */
+export type OwnedById = Brand<string, "OwnedById">;
+/** An account ID of an actor that has updated something */
+export type UpdatedById = Brand<string, "UpdatedById">;
+/** An ID to uniquely identify an account (e.g. a User or an Org) */
+export type AccountId = Brand<string, "AccountId"> | OwnedById | UpdatedById;
+
+/** A Uuuid that points to an Entity without any edition */
+export type EntityUuid = Brand<string, "EntityUuid">;
+/** Valid Uuids of the system */
+export type Uuid = Brand<string, "Uuid"> | AccountId | EntityUuid;
+
+// These type overwrites are centralized for being able to swap out implementations.
+
+export const splitEntityId = splitEntityIdSubGraph as (
+  entityId: EntityId,
+) => [OwnedById, EntityUuid];
+
+export const extractOwnedByIdFromEntityId =
+  extractOwnedByIdFromEntityIdSubGraph as (entityId: EntityId) => OwnedById;
+
+export const extractEntityUuidFromEntityId =
+  extractEntityUuidFromEntityIdSubGraph as (entityId: EntityId) => EntityUuid;
+
+/** If the underlying entityUuid is an accountId, use this cast to convert the type */
+export const extractAccountIdAsEntityUuid =
```

Comment:
I was thinking about this in the discussion we were just having about the hierarchy of branding and if we actually want/need an `OwnedById`.

First of all, I think the naming of this is backwards, it should be `extractEntityUuidAsAccountId` right?

Secondly, I'm not sure this helper method is a good idea, it'd be too easy to call this on the wrong kind of `EntityId` with a "trust me" approach rather than trying to embed further type safety.

What if we had an `EntityId` subtype for `AccountEntities` or something? (maybe a slightly less confusing name would be good), and then we have an `extractAccountId` function that takes in that branded `EntityId`. Then you'd need to explicitly say "the `entityId` I have here is for one of our unique account entities"

Reply:
Yes, I see now that the name is backwards. And I do like that approach instead, I'll look into that :+1:

Reply:
0e0e8ab0b3cdb887fc1f18aeb015bd5630f79675 should be addressed here. Using the `brand` fn directly